broadinstitute/single_cell_portal_core
# Instantly filter cells across annotations: client library (SCP-5213)

This implements a library to let users instantly filter plotted cells across annotations like "cell type", "disease", and more.

### Overview
Previously, users could filter cells only by groups in the annotation that is currently selected.  For example, if viewing the annotation "Infant sick", users could filter the cluster scatter plot to only show cells in (say) the groups "Yes" or "Not applicable" that are part of that single annotation.  

Now, this new client-side cell faceting library enables a UI to let users filter cells across _several_ annotations.  So users can refine the plotted cells to only those that have e.g. "Infant sick: Yes or Not applicable" **and** "Cell type: Epithelial cell" **and** "Daycare: Yes".  As is conventional in faceted search, selections _within_ a facet are Boolean ORs, and selections _across_ facets are Boolean ANDs.

Filtering is now also separable from coloring: you could e.g. color the resulting cluster scatter plot by, say "Donor ID".  This means that cell faceting is only relevant when there are multiple applicable annotations.

### Video

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/e2e6bc13-4723-4501-b33d-e0442c17ce3f

### Performance
It's extremely fast.

Whether for 50K cells or 1.3M cells, this client-side cell filtering takes 30-80 ms on my machine.  My laptop is a 14" 2021 MacBook Pro, with an Apple M1 Pro CPU and 16 GB RAM.  This is faster than most, but likely not abnormally higher-specced than the typical SCP user.  I was also running a screen recording, web server, IDE, and other programs during these local spot measurements.  This fast filtering is powered by [Crossfilter](https://github.com/crossfilter/crossfilter).

### Logging
The cell faceting library has analytics instrumentation which logs performance timing (`perfTime`) and related properties to Mixpanel for a new `filter-cells` log event.

### Test
Automated tests verify behavior in this new code.  If desired, to manually test it:

1. In ExploreDisplayTabs.jsx, uncomment the line `// window.SCP.updateFilteredCells = updateFilteredCells`
2. Go to DE pilot study ("Human milk - differential expression")
3. In the Console panel in DevTools, run `window.SCP.updateFilteredCells({'cell_type__ontology_label--group--study': ['epithelial cell']}) // Apply 1 facet, 1 filter`
4.  Confirm the cluster scatter plot updates to show only cells in the "LC1" and "LC2" groups, with cell counts 4427 and 35398 respectively.
5. In Console, run `window.SCP.updateFilteredCells({'cell_type__ontology_label--group--study': ['epithelial cell'], 'infant_sick_YN--group--study': ['yes', 'NA']}) // Apply 2 facets, 3 filters`.  In human-friendly Boolean terms, this translates to "Cell type: epithelial cell" AND ("Infant sick: yes OR NA").
6. Confirm LC1 cell count updates to 1090, LC2 to 10725.
7. In Console, run `window.SCP.updateFilteredCells({}) // Clear filters`
8. Confirm cluster scatter plots returns to its original state
9. In ExploreDisplayTabs.jsx, comment out the line `window.SCP.updateFilteredCells = updateFilteredCells`

*Be sure you do that last step!*  If you don't, tests will fail.

### Further details
This client-side library uses data from the new `/facets` endpoint described in #1838.  More context is available in this week's [SCP demo recording](https://drive.google.com/file/d/1whqhirP8ciQx4tkKnQ5No8CB9RzyBGbY/view?t=19m20s).  Sometimes, filtering is called "subsetting".  Per stakeholder feedback, one next step here is to use pale grey color / send-to-back for cells omitted by filtering (SCP-5277), rather than simply removing them from the plot.  

This satisfies SCP-5213.
